third_party/move/move-vm/types/src/unit_tests/value_exchange_tests.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    value_exchange::{
+        deserialize_and_exchange, serialize_and_exchange, AsIdentifier, ExchangeResult,
+        IdentifierBuilder, ValueExchange,
+    },
+    values::{Struct, Value},
+};
+use move_core_types::value::{LayoutTag, MoveStructLayout::Runtime, MoveTypeLayout};
+use std::{cell::RefCell, collections::BTreeMap};
+
+#[cfg(test)]
+#[derive(Debug, Default)]
+struct TestExchange {
+    // For testing purposes, all swapped data is stored in a map.
+    liftings: RefCell<BTreeMap<u64, Value>>,
+}
+
+#[cfg(test)]
+impl ValueExchange for TestExchange {
+    fn try_exchange(&self, value_to_exchange: Value) -> ExchangeResult<Value> {
+        // Identifiers are generated using the number of entries stored
+        // so far.
+        let mut liftings = self.liftings.borrow_mut();
+        let identifier = liftings.len() as u64;
+
+        let identifier_value = value_to_exchange.build_identifier(identifier).unwrap();
+        liftings.insert(identifier, value_to_exchange);
+        Ok(identifier_value)
+    }
+
+    fn try_claim_back(&self, value_to_exchange: Value) -> ExchangeResult<Value> {
+        let liftings = self.liftings.borrow();
+
+        let identifier = value_to_exchange.as_identifier().unwrap();
+        Ok(liftings.get(&identifier).unwrap().copy_value().unwrap())
+    }
+}
+
+#[test]
+fn test() {
+    let exchange = TestExchange::default();
+
+    // We should swap 2nd and 3rd fields.
+    let value = Value::struct_(Struct::pack(vec![
+        Value::u64(100),
+        Value::u128(101),
+        Value::u64(102),
+    ]));
+
+    let layout = MoveTypeLayout::Struct(Runtime(vec![
+        MoveTypeLayout::U64,
+        MoveTypeLayout::Tagged(LayoutTag::AggregatorLifting, Box::new(MoveTypeLayout::U128)),
```

Comment:
Can we test mismatch? I.e. if We have layout that says U128 but the Value type isn't correct. Asking because we create a tagged layout as well, and if there is an issue there, we still want to handle in a predictable way.

Reply:
Not sure if that's the same, but let's try to test ExchangeError case

Reply:
Added some tests for non supported cases👍